src/ForwardDiff.jl

```diff
@@ -2,90 +2,106 @@ isdefined(Base, :__precompile__) && __precompile__()
 
 module ForwardDiff
 
-    if VERSION < v"0.4-"
-        warn("ForwardDiff.jl is only officially compatible with Julia v0.4-. You're currently running Julia $VERSION.")
-    end
+using Compat
 
-    ###########
-    # imports #
-    ###########
-    import Calculus
-    import NaNMath
-    import Base: *, /, +, -, ^, getindex, length,
-                 hash, ==, <, <=, isequal, copy, zero,
-                 one, float, rand, convert, promote_rule,
-                 read, write, isless, isreal, isnan,
-                 isfinite, isinf, eps, conj, transpose,
-                 ctranspose, eltype, abs, abs2, start,
-                 next, done, atan2, floor, ceil, trunc, round
-
-    const auto_defined_unary_funcs = map(first, Calculus.symbolic_derivatives_1arg())
-
-    for fsym in auto_defined_unary_funcs
-        @eval import Base.$(fsym);
-    end
+import Calculus
+import NaNMath
+
+#######################
+# compatibility patch #
+#######################
 
-    ############
-    # includes #
-    ############
-    include("Partials.jl")
-    include("ForwardDiffNumber.jl")
-    include("GradientNumber.jl")
-    include("HessianNumber.jl")
-    include("TensorNumber.jl")
-    include("api/api.jl")
-
-    #######################
-    # Promotion Utilities #
-    #######################
-    @inline switch_eltype{T,S}(::Type{Vector{T}}, ::Type{S}) = Vector{S}
-    @inline switch_eltype{N,T,S}(::Type{NTuple{N,T}}, ::Type{S}) = NTuple{N,S}
-
-    for F in (:TensorNumber, :GradientNumber, :HessianNumber)
-        @eval begin
-            @inline switch_eltype{N,T,S}(::Type{$F{N,T,NTuple{N,T}}}, ::Type{S}) = $F{N,S,NTuple{N,S}}
-            @inline switch_eltype{N,T,S}(::Type{$F{N,T,Vector{T}}}, ::Type{S}) = $F{N,S,Vector{S}}
-
-            function promote_rule{N,T1,C1,T2,C2}(::Type{($F){N,T1,C1}},
-                                                 ::Type{($F){N,T2,C2}})
-                P = promote_type(Partials{T1,C1}, Partials{T2,C2})
-                T, C = eltype(P), containtype(P)
-                return ($F){N,T,C}
-            end
-
-            function promote_rule{N,T,C,S<:Real}(::Type{($F){N,T,C}}, ::Type{S})
-                R = promote_type(T, S)
-                return ($F){N,R,switch_eltype(C, R)}
-            end
-        end
+if v"0.4" <= VERSION < v"0.5-"
```

Comment:
please, please, **please** be more specific with VERSION dependent code like this